### PR TITLE
adding reddit listen support

### PR DESCRIPTION
**Pull request by stedn**
Getting near the home stretch hopefully.  This works on the local server past through to the point where it loads the reddit_user.html template.  However, I can't figure out how to launch tasks.py to test the actual pull.
Screenshot_2020-04-16_01-17-07

Also, it turns out in the belly of `models.py -> _urls_and_domains()` there is a call to granary's `gr_source.user_to_actor` that passes the original user_json from oauth-dropins! So I've added a compatibility change to oauth-dropins (https://github.com/snarfed/oauth-dropins/pull/35) and granary (https://github.com/snarfed/granary/pull/200) that now converts from praw to an intermediate user_dict.

Needless to say there will need to be some tests here too, but I wanted to get feedback on how tasks.py should be set off to check that part.

### CODE REVIEW

File: templates/about.html [hunk 1]
```diff
@@ -327,6 +327,7 @@
...
   <li><b>Twitter</b> <a href="https://developer.twitter.com/en/docs/basics/authentication/faq">distinguishes between read and write access,</a> but not between different types of data. We ask for read access to your tweets and profile for listen, write access for publish. We don't ask for access to your direct messages. Screenshots: <a href="/static/permissions_screenshots/twitter_read.png">listen</a>, <a href="/static/permissions_screenshots/twitter_write.png">publish</a>.</li>
   <li><b>WordPress.com</b>: doesn't have permissions at all, <a href="https://developer.wordpress.com/docs/oauth2/">just full access or nothing.</a> We ask for read and write access to a single blog that you specify, so we can see and handle new blog posts and to create comments. <a href="/static/permissions_screenshots/wordpress.com.png">Screenshot.</a></li>
+  <li><b>Reddit</b> allows you to permit access to different <a href="https://www.reddit.com/api/v1/scopes.json">scopes</a>, We're asking for permission to confirm your identity and read posts that you can read.</li>
```
Comment (snarfed) [inline templates/about.html:330]: minor typo nit, lower case "we."

File: reddit.py [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+  The key name is the username.
+  """
+  GR_CLASS = gr_reddit.Reddit
+  OAUTH_START_HANDLER = oauth_reddit.StartHandler
+  SHORT_NAME = 'reddit'
+  TYPE_LABELS = {
+    'post': 'submission',
+    'comment': 'comment',
+  }
+  TRANSIENT_ERROR_HTTP_CODES = ('404',)
```
Comment (snarfed) [inline reddit.py:29]: oh wow, reddit 404s are transient?! that's...odd and unfortunate.

File: reddit.py [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+
+  @staticmethod
+  def new(handler, auth_entity=None, **kwargs):
+    """Creates and returns a :class:`Twitter` entity.
```
Comment (snarfed) [inline reddit.py:33]: s/Twitter/Reddit/, here and below

File: reddit.py [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+    return self.gr_source.user_url(self.key_id())
+
+  def label_name(self):
+    """Returns the username."""
+    return self.key_id()
+
+  def search_for_links(self):
+    """Searches for activities with links to any of this source's web sites.
+
+    Twitter search supports OR:
```
Comment (snarfed) [inline reddit.py:60]: remove these two paragraphs? feel free to update with reddit-specific details if necessary, otherwise omit

File: reddit.py [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+
+    ...but it only returns complete(ish) results if we strip scheme from URLs,
+    ie search for example.com instead of http://example.com/, and that also
+    returns false positivies, so we check that the returned tweets actually have
+    matching links. https://github.com/snarfed/bridgy/issues/565
+
+    Returns:
+      sequence of ActivityStreams activity dicts
+    """
+    urls = set(util.schemeless(util.fragmentless(url), slashes=False)
```
Comment (snarfed) [inline reddit.py:71]: does reddit search also need the scheme removed from URLs? otherwise try the full URLs, like normal?

File: reddit.py [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+    if not urls:
+      return []
+
+    candidates = []
+    for u in urls:
+      candidates.extend(self.get_activities(
+        search_query=query, group_id=gr_source.SEARCH, etag=self.last_activities_etag,
+        fetch_replies=False, fetch_likes=False, fetch_shares=False, count=50))
+
+    # filter out retweets and search false positives that don't actually link to us
```
Comment (snarfed) [inline reddit.py:83]: remove this block entirely?

File: reddit.py [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+    handler = util.oauth_starter(oauth_reddit.StartHandler, feature=feature).to(
+      '/reddit/add')(self.request, self.response)
+    return handler.post()
+
+
+class AddReddit(oauth_reddit.CallbackHandler, util.Handler):
+  def finish(self, auth_entity, state=None):
+    logging.debug('finish with %s, %s', auth_entity, state)
+    self.maybe_add_or_delete_source(Reddit, auth_entity, util.encode_oauth_state(state))
+
```
Comment (snarfed) [inline reddit.py:123]: very minor style nit, one more newline

File: reddit.py [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+class StartHandler(AuthHandler):
+  """Custom OAuth start handler so we can use access_type=read for state=listen.
+
+  Tweepy converts access_type to x_auth_access_type for Twitter's
```
Comment (snarfed) [inline reddit.py:127]: minor nit, remove

**Comment by snarfed** [comment]
re running the poll task, sorry, the port will actually be 8081, or whatever port dev_appserver.py said it started the "background" module on. so something like:

> `curl -v -d 'source_key=TODO&last_polled=0' http://localhost:8081/_ah/queue/poll`

**Comment by snarfed** [comment]
i'm actually not having much luck with that myself. seems to hang, then 503, with no logging from dev_appserver.

ugh. sorry.

**Comment by stedn** [comment]
thanks, clearly I did not even look at the comments.  oops.  i'll try toying around more tonight and see if I can get tasks.py running. thanks for looking into it.

File: reddit.py (hunk 2 above)
Comment (stedn) [reply reddit.py]: oh, this is probably not the case.  this was just a copy paste from twitter.py, and I didn't know what it did so I left it.  I've removed it now and it still seems to work.

File: reddit.py (hunk 5 above)
Comment (stedn) [reply reddit.py:62]: i'm honestly not sure what scheme means here. but if you mean the http part I think it needs to be removed.

Comment (snarfed) [reply reddit.py:62]: yeah scheme is protocol. you can probably test searches for different URL formats in reddit's UI if you want.

**Comment by snarfed** [comment]
figured out running a task. i'm adding this to the readme:

To test polling interactively, temporarily comment out the `manual_scaling` section in [background.yaml](https://github.com/snarfed/bridgy/blob/master/background.yaml), uncomment the `automatic_scaling` section, and use `curl` to run the task handler directly:

```sh
curl -d 'source_key=[KEY]&last_polled=0' http://localhost:8081/_ah/queue/poll
```

...where `[KEY]` is the url-safe ndb key for the source entity you want to poll, e.g. `ndb.Key('Twitter', 'schnarfed').urlsafe()`.

File: reddit.py (hunk 5 above)
Comment (stedn) [reply reddit.py:62]: with http in the search, it doesn't find it, so I'll use the schemeless.

**Comment by stedn** [comment]
Along with two more minor changes in oauth-dropins https://github.com/snarfed/oauth-dropins/pull/36 and granary https://github.com/snarfed/granary/pull/201 this now results in a working poll that attempts to send the webmentions.  
image

Apologies for the multiple revisions, but hopefully this is still easier than working through three finished mega-PRs at once.

Currently the webmentions aren't actually arriving at my domain (afaict), but I assume I'm missing another step that has to be triggered asynchronously?  I have to admit that tracing the code flow through all of these callbacks is not my specialty.

**Comment by snarfed** [comment]
exciting!

the webmentions aren't making it all the way to your site because the source URLs are on localhost:8080, on your local bridgy, which your site can't fetch.

you can fetch the local mf2 urls to see how they look though! see https://brid.gy/about#source-urls

**Comment by snarfed** [comment]
oh the other reason is that sending webmentions are their own tasks that you have to trigger manually too. 😐

actually now that i look, the easiest way to construct the curl command to trigger tasks is to look for dev_appserver logs starting with _Would add task: ..._ and put the request body there directly into `curl -d ...`. eg for a webmention task:

```sh
curl -v -d 'response_key=agNhcHByVAsSCFJlc3BvbnNlIkZ0YWc6cGljb25pYy5jbywyMDEzOjE1NzE3ODk1NDY2NTU2MjExMl9yZWJsb2dnZWRfYnlfMTQ1Mzg0NzYxNDI1MjY4NzM2DA' localhost:8081/_ah/queue/propagate
```

**Comment by snarfed** [comment]
...and then, even if that actual webmention fails, you can find bridgy's mf2 handler URL in the _Webmention from .... to ...` log line and fetch it to see if it looks ok. eg http://localhost:8081/repost/pixelfed/@snarfed@piconic.co/157178954665562112/145384761425268736

**Comment by stedn** [comment]
ok, wow, it took me forever to figure out how to make that happen, but I can see it.  It redirects after just a second though (but I assume that is normal).  
image

I needed to make one more addition to granary to add the `get_comments()` function.  I will add to the current PR (https://github.com/snarfed/granary/pull/201) that is open.

File: tests/test_reddit.py [hunk 9]
```diff
@@ -0,0 +1,47 @@
+"""Unit tests for twitter.py.
```
Comment (snarfed) [inline tests/test_reddit.py:1]: reddit

File: tests/test_reddit.py [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+import urllib.request, urllib.parse, urllib.error
+
+from granary import reddit as gr_reddit
+from granary.tests import test_reddit as gr_reddit_test
+import oauth_dropins.reddit
+from oauth_dropins.webutil.util import json_dumps, json_loads
+
+import models
+from . import testutil
+import reddit
```
Comment (snarfed) [inline tests/test_reddit.py:13]: unused

File: tests/test_reddit.py [hunk 11]
```diff
@@ -0,0 +1,47 @@
+"""Unit tests for twitter.py.
+"""
+import copy
+import urllib.request, urllib.parse, urllib.error
+
+from granary import reddit as gr_reddit
```
Comment (snarfed) [inline tests/test_reddit.py:6]: unused

File: tests/test_reddit.py [hunk 12]
```diff
@@ -0,0 +1,47 @@
+"""Unit tests for twitter.py.
+"""
+import copy
```
Comment (snarfed) [inline tests/test_reddit.py:3]: unused

File: tests/test_reddit.py [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+import urllib.request, urllib.parse, urllib.error
```
Comment (snarfed) [inline tests/test_reddit.py:4]: unused

File: tests/test_reddit.py (hunk 9 above)
Comment (stedn) [reply tests/test_reddit.py]: ah! I was so close on the doc strings this time.

**Comment by stedn** [comment]
I deployed and it is working for me!

image

I'll add something to the README about deploying to your own app engine instance.  There are only a couple of tweaks that I needed to make it work.

I don't have my own keys for twitter etc yet, but I've applied and am waiting. I could run my own mirror to take load off of bridgy central.  I'll see if I stay below the free limit (and if not I can tweak it to poll less frequently).  maybe call it step.py stones (as opposed to a proper brid.gy, ok clearly tired not makin sense anymore).

**Comment by snarfed** [comment]
> I deployed and it is working for me!

oh wow, awesome! nice work!

> I could run my own mirror to take load off of bridgy central.

thanks for the thought! let's just stick with the main one though. load is relatively even from all backfeed users, though, and bridgy proper currently has >4k of those, so one more won't make a difference, and i very much want to avoid fragmentation across multiple instances, which would make support and maintenance a lot harder.

deploying your own instance can definitely still be useful for testing though! mind updating the new instructions to say it's ok for testing, but discouraged for production use?

**Comment by stedn** [comment]
OK, I added that sentence.  But you should consider that as an option someday.  Seems like everybody's doing the decentralized instance thing these days.

File: reddit.py [hunk 14]
```diff
@@ -0,0 +1,115 @@
...
+        fetch_replies=True, fetch_likes=False, fetch_shares=False, count=50))
+
+    return candidates
+
+
+class AuthHandler(util.Handler):
+  """Base OAuth handler class."""
+
+  def start_oauth_flow(self, feature):
+    """Redirects to Twitter's OAuth endpoint to start the OAuth flow.
```
Comment (snarfed) [inline reddit.py:81]: minor nit: Reddit

**Comment by snarfed** [comment]
decentralized! federated! for services that manage user data, yeah, definitely interesting. bridgy is a pure service though, not a user data repository, so the benefits are a bit less obvious.

regardless. looking good! my only last thought is that granary reddit doesn't have preview/create yet, so bridgy reddit can't actually support publish just yet. want to remove that in this PR for now? you can use instagram as an example of a source that only supports listen.

File: reddit.py [hunk 15]
```diff
@@ -0,0 +1,112 @@
...
+  The key name is the username.
+  """
+  GR_CLASS = gr_reddit.Reddit
+  OAUTH_START_HANDLER = oauth_reddit.StartHandler
+  SHORT_NAME = 'reddit'
+  TYPE_LABELS = {
+    'post': 'submission',
+    'comment': 'comment',
+  }
+
```
Comment (snarfed) [inline reddit.py:29]: to disable publish, i think all you need to do is add `CAN_PUBLISH = False` here.

Comment (stedn) [reply reddit.py:30]: ah cool, i missed that.

**Comment by stedn** [comment]
yes, i ran through signup, disable, and signup again on my deployed version and it seems to work.

**Comment by snarfed** [comment]
i did too just now, worked great!

**Pull request merged**